### PR TITLE
RFC Process Update

### PR DESCRIPTION
**Pull request by locks**
[Rendered](https://github.com/emberjs/rfcs/blob/rfc-process-update/text/0300-rfc-process-update.md)

### CODE REVIEW

File: text/0000-rfc-process-update.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+## Motivation
+
+Inspired by Rust, Ember implemented a Request For Commits (RFC) process with the intent to involve the community in discussing the future of the framework.
+The RFC process has proved invaluable.
+Proposals by both Core and community members are discussed and refined,
+with the result coming out much stronger.
+Other projects like React also adopted an RFC process, lending external validation to this idea.
+
+After having been involved with the RFC process for some time, and from feedback from the community, some issues with the current implementation of the process have been noted.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Following the success of its initial implementation, the RFC process can be further refined to address some issues identified by the community and core team members.

File: text/0000-rfc-process-update.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+### Confusion about emberjs/rfcs and ember-cli/rfcs
+
+The Ember project currently has two separate RFC processes for Ember.js and Ember CLI.
+
+This leads to confusion due to having to keep track of two different repositories to track progress.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: This leads to confusion because users must keep track of two different repositories in order to monitor progress.

File: text/0000-rfc-process-update.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+It also increases the burden on the submitter of an RFC, as the two processes have diverged slightly,
+making the submitter be aware of these differences during the lifecycle of the RFC.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: The submitter must consider these differences during the lifecycle of the RFC.

File: text/0000-rfc-process-update.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+There are also some cases where it is not clear exactly which repository the RFC should be submitted to.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: exactly where the RFC should be submitted.

File: text/0000-rfc-process-update.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+Once again taking inspiration from Rust, they have experimented expanding the types of RFCs proposed
+[crates.io default ranking](https://github.com/rust-lang/rfcs/blob/master/text/1824-crates.io-default-ranking.md)
+[default URLs for documentation](https://github.com/rust-lang/rfcs/blob/master/text/1826-change-doc-default-urls.md).
+The latter one, for example, would have been a useful RFC when designing the new [API docs](https://github.com/ember-learn/ember-api-docs).
+
+Rust has RFCs for documentation, website, internals, tools.
+We only have emberjs and ember-cli due to them having been implemented separately.
+
+This makes it so that a potential contributor does not know where to submit a proposal.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: drop, moved up higher

File: text/0000-rfc-process-update.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+### Lingering RFCs
+
+The emberjs/rfcs repository currently has [54 issues](https://github.com/emberjs/rfcs/issues) and [62 PRs](https://github.com/emberjs/rfcs/pulls) open, and ember-cli/rfcs has [183 issues](https://github.com/ember-cli/ember-cli/issues) and [49 PRs](https://github.com/ember-cli/ember-cli/pulls).
+A significant percentage of these have not been active in the recent past.
+
+We have kept PRs and issues open so people could more easily find the discussions,
+but this has instead given a negative feeling of staleness, as RFCs linger open without new feedback.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Maybe give an example or cite a source

File: text/0000-rfc-process-update.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+## Detailed design
+
+To address the issues raised in the “Motivation” section, I will propose changes to the current process.
+
+### Implement Core Champion
+
+To make sure that RFCs receive adequate support from the team, Ember CLI has implemented the idea of a champion associated with each RFC.
+The goal is that in seeking a champion from the team,
+the RFC author starts a dialogue with the team and gets some early feedback.
+The champion is then responsible for representing the RFC in team meetings, and for shepherding its progress.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: no comma

File: text/0000-rfc-process-update.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+This will be imported into emberjs/rfcs.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: This champion model will be adopted for emberjs/rfcs.

Comment (jenweber) [reply text/0000-rfc-process-update.md]: again, maybe link to an example in ember-cli

File: text/0000-rfc-process-update.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+### Improve the lifecycle of RFCs
+
+To address the triaging and inactivity problem of RFCs, I introduce the concept of FCP to close.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: should say "concept of Final Comment Period (FCP) to close"

File: text/0000-rfc-process-update.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+Closing an RFC should be viewed as another triaging tool, not as a rejection of the RFC.
+Sometimes a rewrite of an RFC would be so fundamental that it would benefit of a fresh discussion in a new thread.
+Sometimes the original author is no longer active (Champions should help here as well),
+and someone else might want to take over the work in a new RFC.
+Sometimes the timing might not be right, or the feature might have been addressed some other way, and yes,
+sometimes it might be something that is not aligned with the team's values for the project.
+
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Make the above a bulleted list

File: text/0000-rfc-process-update.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+A good example of this is the [Named Blocks RFC](https://github.com/emberjs/rfcs/blob/master/text/0226-named-blocks.md#motivation),
+which lists in the motivation section previous attempts at similar ideas.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: which covers previous attempts in the "Motivation" section.

File: text/0000-rfc-process-update.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+Like the FCP to merge process, once an RFC is marked as FCP to close there will be a period of one week where people can raise new concerns.
+After that period of one week, the respective team will review and close the RFC or extend the period for another week.
+
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: I think this is the most likely section people may object to. Try to anticipate the objections and work them in here. i.e. "won't this discourage submitters from participating in the future?"

File: text/0000-rfc-process-update.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+### Merge ember-cli/rfcs into emberjs/rfcs
+
+Part of removing confusion about the process is having a central repository for all Ember RFCs.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: There will be a central repository for all Ember RFCs.

File: text/0000-rfc-process-update.md [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+To achieve merging ember-cli/rfcs into emberjs/rfcs the following will be done:
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Here is how ember-cli/rfcs can be merged into emberjs/rfcs:

File: text/0000-rfc-process-update.md [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+* Remove active/completed concepts from ember-cli/rfcs
+  * Add front matter
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: I don't know what this is, but if other people do, you could keep as-is

File: text/0000-rfc-process-update.md (hunk 7 above)
Comment (jenweber) [reply text/0000-rfc-process-update.md]: Also as a user, I would want to see formal affirmation that a decent number of core team members _would_ champion other RFCs before I'd feel good about this section. If this falls to just a few people, it will not work and we'll burn out them out

File: text/0000-rfc-process-update.md [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+  * Copy both active and completed RFC files into `text` of emberjs/rfcs
+* Transfer active PRs and Issues to emberjs/rfcs
+* Archive ember-cli/rfcs
+
+There are some concerns about links breaking when we move the files to emberjs/rfcs,
+but given the fact that ember-cli/rfcs had the concept of active/completed by moving the files into different folders,
+links were already being broken.
+
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Phrase more positively - It is expected that some links will break in the process of moving files to emberjs/rfcs. This does not introduce any new concerns, since ember-cli/rfcs already had a process to move files into active/completed folders, which also breaks the links.

File: text/0000-rfc-process-update.md [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+the RFC front matter should be oriented towards those teams.
+
+A list of the teams and respective projects should be added to the instructions,
+possibly with the addition of per-team instructions on specifics of the project.
+Additional templates might be created as well, such a design work template.
+
+### Provide more guidance for post-RFC
+
+At the moment it is not clear for some contributors what happens once an RFC is merged.
+How and when should work to implement an RFC be done?
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Needs another sentence or two. Maybe just specifying that the instructions will be added to a CONTRIBUTING.md file or something. It would be good to see a draft and link it here.

File: text/0000-rfc-process-update.md [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+## How We Teach This
+
+To ensure that contributors are updated on the RFC process and the process is clear,
+the documentation should be improved in a couple of ways.
+
+The README should be updated to include information about Champions,
+how they are assigned, what they do.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: and what they do

File: text/0000-rfc-process-update.md [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+
+## Drawbacks
+
+### Adjustment period
+
+There are active RFCs in ember-cli/rfcs. Moving these discussions would be onerous, so they should be kept there until completion, and no new RFCs accepted.
+
+### Permalinks to ember-cli/rfcs proposals
+
+Moving the RFC files from ember-cli/rfcs (active or completed) to emberjs/rfcs can be seen as a breaking change, and could lead to someone linking to ember-cli/rfcs and then the RFC being updated in emberjs/rfcs. However, ember-cli/rfcs already suffers from a linking problem due to the active/completed folders, as RFCs need to be moved from one to the other even after being accepted.
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Repetition with "There are some concerns about links breaking..." I recommend removing the "There are concerns" section and working it into here

File: text/0000-rfc-process-update.md [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+None at the moment.
+
+## Unresolved questions
+
+### How do Champions get assigned to an RFC?
+
+More enterprising authors will likely reach out to team members on their own,
+but the added burden of having to seek a Champion might discourage people from authoring RFCs.
+
+The various teams could agree to review new RFCs during the weekly calls and assign Champions?
```
Comment (jenweber) [inline text/0000-rfc-process-update.md]: Maybe list off a few more ways to do this, in a bulleted list, and drop the "?"

File: text/0000-rfc-process-update.md (hunk 7 above)
Comment (jenweber) [reply text/0000-rfc-process-update.md]: It's kind of a side comment. I'm hoping that team members will chime in and say, "yes I want to do this," not necessarily that people should be listed here

File: text/0000-rfc-process-update.md [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+# RFC Process Update
+
+## Summary
+
+Introduce the idea of Core Champion to the Ember.js RFC process, merge Ember.js and Ember CLI RFCs repositories,
+and add metadata to make cross-project RFCs easier.
+
+## Motivation
+
+Inspired by Rust, Ember implemented a Request For Commits (RFC) process with the intent to involve the community in discussing the future of the framework.
```
Comment (ro0gr) [inline text/0000-rfc-process-update.md]: I'm confused about RFC abbreviation. I've heard at least about three interpretations:
  - request for **change**
  - request for **comments**
  - and now request for **commits**

Does anybody knows for sure what does "C" stand for?

Comment (locks) [reply text/0000-rfc-process-update.md]: It's a typo, should be Request For Comments [[1]](https://github.com/emberjs/rfcs#ember-rfcs) [[2]](https://github.com/rust-lang/rfcs/#rust-rfcs).

**Comment by Gaurav0** [comment]
Thanks for this. I have one additional suggestion: a standard time period after which an FCP to close should be expected. Perhaps 6 months. The idea should be that RFCs should not be perceived to be treated differently based on author and all RFCs should have sufficient time for comment, no matter how unpopular.

**Comment by Gaurav0** [comment]
One other concern: there should not be a rather of FCP to close immediately after this is accepted.

File: text/0000-rfc-process-update.md (hunk 2 above)
Comment (sandstrom) [reply text/0000-rfc-process-update.md]: How about `people/participants/contributors` over `users`? 😄

**Comment by joukevandermaas** [comment]
The part about assigning champions is not entirely clear to me. Does the RFC author seek out a champion? Or do they just submit the RFC and wait for one to assign themselves?

If the former, it may help to have some semi-standardized way of contacting core team members for this purpose (perhaps a slack channel). This should be mentioned somewhere in the repo so people know where to start.

If the latter, it may be a good idea to have some time-frame in which a decision is made and communicated (even if the decision is that no one has time to champion that RFC for now). My ballpark suggestion would be about 6 weeks after opening the RFC (so about one release cycle). I know time-frames are difficult to commit to, so maybe this is impossible.

**Comment by rwjblue** [comment]
> The part about assigning champions is not entirely clear to me. Does the RFC author seek out a champion? Or do they just submit the RFC and wait for one to assign themselves?

I think that in practice, interested team members are _falling over themselves_ to champion ideas that they like and/or want to advance.  I know that I do... But I think the answer here is actually "both". Advocating for the RFC both in its own comments, in `#dev-rfcs` in slack, on the forum, etc is likely the best way...

> If the latter, it may be a good idea to have some time-frame in which a decision is made and communicated (even if the decision is that no one has time to champion that RFC for now). My ballpark suggestion would be about 6 weeks after opening the RFC (so about one release cycle). I know time-frames are difficult to commit to, so maybe this is impossible.

I'm not sure this is a great idea. I don't think we need more process just for process sake, all these rules are just one more thing that we have to maintain/adhere to/etc.  I definitely think that a nice default timeframe to `FCP to close` after not successfully engaging a champion is warranted (I believe that @Gaurav0 also suggested this).

**Comment by rwjblue** [comment]
I'd love to see this move forward, @locks are you still planning on pushing this?

**Comment by mehulkar** [comment]
Is it possible to move ember-cli into the emberjs org as well? It's a core part of the Ember experience now and IMO if RFCs are merged, it should be located closer to the library also. IMO, many of the other repos on ember-cli could also be moved into `github.com/emberjs` or maybe github.com/emberjs-core-addons` or something similar.

**Comment by ctcpip** [comment]
> Each RFC will require a champion from the primary core team to which the RFC has been marked relevant. The champion must be found by the opener of the RFC or other community member.

So basically if a core member can't be convinced to champion, the RFC is dead in the water?  This does not seem community-driven to me.  It sounds like an oligarchy.  I understand the motivation behind it, but it seems like it's taken a step too far.

**Comment by ctcpip** [comment]
@MelSumner -- I don't mean to be pedantic, but that section could use some work in terms of elaborating on how exactly the required champion manifests.  Yes the core team is community, but is there any concern that as their numbers are limited, they may be reluctant to champion otherwise-worthy RFCs because they are already very busy?  I may be interpreting the intent too strictly, but if so, I think that section could be given a bit more detail. 

> "A section on 'Finding a champion' will be added to the instructions on proposing an RFC."

Maybe that would have more specifics?  The language just seems to be a bit discouraging to me and implies that without a champion, you may as well not submit an RFC.  For example, in some projects, if you miss some teeny little instruction when submitting an issue, the maintainer will close it outright, even though it's a valid issue.  I'm not saying that's analogous to what's happening here, but from an outsider looking in, it may look similar.  As I understand it, this comes at a time when increasing community involvement has been a priority, so some sensitivity around something like this I think is warranted.

To put it another way, I'm not cynical about the intent of the RFC or of the folks proposing it.  I'm concerned about the language (and maybe part of the process itself?) possibly discouraging contribution by implying (or seeming to imply) that this additional requirement be met before being heard.

**Comment by kategengler** [comment]
The Champion requirement has been in place on ember-cli/rfcs since early 2017. It was implemented to attempt to prevent RFCs from stagnating. It has served ember-cli well. 

The current reality is that if an RFC does not have a core team member advocating for it -- making sure the core team(s) read it, ensuring all RFC comments are addressed, gathering consensus for FCP, for merging -- is that the RFC stays open and stagnates. That's not to say the core team doesn't pay attention to new RFCs (we certainly do), but that if an RFC doesn't entice someone on the team to push it forward, members may read it and forget it. The champion requirement makes this an explicit requirement and provides a guide for fulfilling it. It reflects that RFCs that are most likely to be accepted are ones for which the author "lobbies" -- by sharing ideas in discord, chatting on the forums, giving talks, writing, etc. 

Mel is correct when she says "assume the best". While the numbers of core team members are limited, it is certainly not small: https://emberjs.com/team/ and in practice, team members will be excited to champion any RFC they think improves Ember. Getting a core team member on board to champion would be a reasonable metric to decide whether it is worth spending the time writing an RFC.

**Comment by Gaurav0** [comment]
> So basically if a core member can't be convinced to champion, the RFC is dead in the water? This does not seem community-driven to me. It sounds like an oligarchy. I understand the motivation behind it, but it seems like it's taken a step too far.

The final decision is in the hands of the core team anyway. The RFC Process is about soliciting community input, not creating some sort of democracy. At least Ember's core team is relatively large and comprises many different points of view, and they have proven that they generally don't ignore community needs.

**Comment by mehulkar** [comment]
> Everything is possible 😈 but I don't personally think this would add much value. It is quite nice to have a grouping of packages that relate to "the CLI tool"...

Some counter points:

- Considering that it’s basically impossible to use Ember without CLI (no mention in guides), I think the value add is that users don’t need to hunt down another org. 
- The website is merged, the RFCs are merged, it seems weird for the code not to live in the same place. 
- GH org’s repo search and tags are a pretty decent way to organize / find repos, IMO. 
- in the (very) long term, I expect that much of ember tooling will converge with non-ember tooling, so the need for supporting CLI packages will go down.

But that said, I haven’t vetted all the repos in ember-CLI yet so it’s possible I’m missing something.
